lib/runner/variant-reassignment.js

```diff
@@ -51,79 +73,96 @@ export default class VariantReassignment {
       = this._selectProductDraftsForReassignment(productDrafts, products)
 
     this.logger.debug(
-      'Filtered %d productDrafts for reassignment',
-      productDraftsForReassignment.length
+      `Filtered ${productDraftsForReassignment.length} productDrafts for reassignment`
     )
 
-    for (const productDraft of productDraftsForReassignment)
+    for (const productDraft of productDraftsForReassignment) {
+      const skus = this.productService.getProductDraftSkus(productDraft)
       try {
+        this.statistics.processed++
+        this.statistics.processedSkus.push(...skus)
         await this._processProductDraft(productDraft, products)
+        this.statistics.succeeded++
       } catch (e) {
-        await this._handleProcessingError(productDraft, products, e)
+        failedSkus.push(...skus)
+        await this._handleUnrecoverableError(e)
       } finally {
         this.logger.debug(
-          'Finished processing of productDraft with name %j',
-          productDraft.name
+          `Finished processing of productDraft with name ${JSON.stringify(productDraft.name)}`
         )
       }
+    }
 
-    return Promise.resolve()
+    this.statistics.failedSkus.push(...failedSkus)
+    return {
+      failedSkus,
+      statistics: this.statistics
+    }
   }
 
-  /**
-   * For 400 or 404 errors, we don't repeat the actions as the request itself is wrong.
-   * For other errors, we retry the actions.
-   *
-   * @see https://github.com/commercetools/commercetools-node-variant-reassignment/issues/60
-   */
-  async _handleProcessingError (productDraft, products, error) {
+  async _handleUnfinishedTransactionsError (retryCount) {
+    this.logger.warn(`Retrying(${retryCount}) processing of unfinished transactions`)
+    this.statistics.retries++
+    await this._processUnfinishedTransactions(null, ++retryCount)
+  }
+
+  async _retryProcessingOfProductDraft (productDraft, products, retryCount) {
+    this.logger.warn(`Retrying(${retryCount}) processing of productDraft`, productDraft)
+    this.statistics.retries++
     const transactions = await this.transactionService.getTransactions()
-    if ([400, 404].includes(error.statusCode)) {
-      this.logger.info('Unrecoverable error, will delete backup custom object.')
+    const failedTransaction = transactions.find(({ value }) =>
+      _.isEqual(value.newProductDraft.slug, productDraft.slug)
+    )
+
+    if (failedTransaction)
+    // transaction was created but not finished, try to finish it
+      await this._processUnfinishedTransactions(transactions, ++retryCount)
+    // transaction was not created, try to process productDraft again
+    return this._processProductDraft(productDraft, products, ++retryCount)
+  }
+
+  async _handleUnrecoverableError (error) {
+    error[constants.ERROR_CONTEXT] = 'Unrecoverable error, will delete backup custom object.'
```

Comment:
should we at least log out this custom object? what do you think?

Reply:
https://github.com/commercetools/commercetools-node-variant-reassignment/pull/48/commits/3f97eaca9c18c433ec48d2eddd5441e0e83f6603